hdk/assembly/emulator.py

```diff
@@ -0,0 +1,99 @@
+from array import array
+from collections.abc import Callable, Iterable
+
+from hdk.assembly.code import normalize_instructions
+from hdk.assembly.syntax import AInstruction, Command, Instruction
+
+_JUMP_TABLE: dict[str | None, Callable[[int], bool]] = {
+    None: lambda comp_result: False,
+    "JMP": lambda comp_result: True,
+    "JLE": lambda comp_result: comp_result <= 0,
+    "JNE": lambda comp_result: comp_result != 0,
+    "JLT": lambda comp_result: comp_result < 0,
+    "JGE": lambda comp_result: comp_result >= 0,
+    "JEQ": lambda comp_result: comp_result == 0,
+    "JGT": lambda comp_result: comp_result > 0,
+}
+
+
+_COMP_TABLE: dict[str, Callable[[int, int, int], int]] = {
+    "0": lambda a, d, m: 0,
+    "1": lambda a, d, m: 1,
+    "-1": lambda a, d, m: -1,
+    "D": lambda a, d, m: d,
+    "A": lambda a, d, m: a,
+    "!D": lambda a, d, m: ~d,
+    "!A": lambda a, d, m: ~a,
+    "-D": lambda a, d, m: -d,
+    "-A": lambda a, d, m: -a,
+    "D+1": lambda a, d, m: d + 1,
+    "A+1": lambda a, d, m: a + 1,
+    "D-1": lambda a, d, m: d - 1,
+    "A-1": lambda a, d, m: a - 1,
+    "D+A": lambda a, d, m: d + a,
+    "D-A": lambda a, d, m: d - a,
+    "A-D": lambda a, d, m: a - d,
+    "D&A": lambda a, d, m: d & a,
+    "D|A": lambda a, d, m: d | a,
+    "M": lambda a, d, m: m,
+    "!M": lambda a, d, m: ~m,
+    "-M": lambda a, d, m: -m,
+    "M+1": lambda a, d, m: m + 1,
+    "M-1": lambda a, d, m: m - 1,
+    "D+M": lambda a, d, m: d + m,
+    "D-M": lambda a, d, m: d - m,
+    "M-D": lambda a, d, m: m - d,
+    "D&M": lambda a, d, m: d & m,
+    "D|M": lambda a, d, m: d | m,
+}
+
+
+def run(instructions: Iterable[Instruction], steps: int, memory: array):
+    commands = normalize_instructions(instructions)
+    pc, a, d = 0, 0, 0
+    for _ in range(steps):
+        curr_command: Command = commands[pc]
+        if isinstance(curr_command, AInstruction):
+            a = int(curr_command.symbol)
+            pc += 1
+            continue
+        m = memory[a] if a <= 24576 else None
```

Comment:
Add a global constant HACK_RAM_SIZE instead of hardcoded 24576 value.

Reply:
Instead of assigning the None value to m do the following. If "M" is not in curr_command.comp, assign m value zero. Else if a > 24576 then raise IndexError exception with information about the error.